### PR TITLE
[SPARK-15340][SQL]Limit the size of the map used to cache JobConfs to void OOM

### PR DESCRIPTION
**Pull request by pzzs**
# What changes were proposed in this pull request?

limit the size of the map used to cache JobConfs in SparkEnv.scala
# How was this patch tested?

Run tpcds(orc),  the num of JobConf in dump file < 1000 and OOM not happened

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 1]
```diff
@@ -363,7 +363,7 @@ private[spark] object HadoopRDD extends Logging {
 
   def containsCachedMetadata(key: String): Boolean = SparkEnv.get.hadoopJobMetadata.containsKey(key)
 
-  private def putCachedMetadata(key: String, value: Any): Unit =
+  private def putCachedMetadata(key: String, value: Object): Unit =
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: Could I ask why this should be `Object`?

Comment (DoingDone9) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: CacheBuilder.newBuilder()  return CacheBuilder<`Object`, `Object`>,  i can only get map[String, `Object`]  @HyukjinKwon

**Comment by akohli** [comment]
curious on a couple of things, firstly, you are using CacheBuilder with a cache size of 1000, is that sufficient?  Wouldn't it be better to catch an exception or detect an error when a jobconf is added to the original map; alternatively is this masking an issue that proper GC is not happening for unreferenced cached configs?

**Comment by DoingDone9** [comment]
It's ok. If the size > 1000, the new values will replace the old values. @akohli

**Comment by akohli** [comment]
sure (i get that) but 1000 is an arbitrary number that may or may not sufficient.  why not 10?  why not 10000?

**Comment by DoingDone9** [comment]
I refer to other codes using the maximumSize method  @akohli 
```
/** A cache of Spark SQL data source tables that have been accessed. */
  protected[hive] val cachedDataSourceTables: LoadingCache[QualifiedTableName, LogicalPlan] = {
    val cacheLoader = new CacheLoader[QualifiedTableName, LogicalPlan]() {
      override def load(in: QualifiedTableName): LogicalPlan = {
        logDebug(s"Creating new cached data source for $in")
        getDatasourceRelation(in)
      }
    }

    CacheBuilder.newBuilder().maximumSize(1000).build(cacheLoader)
  }
```

**Comment by DoingDone9** [comment]
cc @yhuai  @srowen

**Comment by rxin** [comment]
Do you know why it OOMs when it is softValues?

**Comment by DoingDone9** [comment]
I'm confused about it too

**Comment by rxin** [comment]
Does it work for you when you changed it to 10000 rather than 1000?

**Comment by DoingDone9** [comment]
The size of JobConf is about 124k  and the memory of my dirver is 10g (> 124k * 10000 = 1.18g), so it is work. @rxin

**Comment by cloud-fan** [comment]
retest this please

**Pull request closed without merging**